src/content/contribute/writing-a-loader.md

```diff
@@ -48,93 +39,111 @@ resolveLoader: {
 
-### Do not keep state between runs and modules
+- `jade-loader`: Convert template to a module that exports a function.
+- `apply-loader`: Executes the function with loader options and returns raw HTML.
+- `html-loader`: Accepts HTML and outputs a valid JavaScript module.
 
-A loader should be independent of other modules compiled (except of these issued by the loader).
+T> The fact that loaders can be chained also means they don't necessarily have to output JavaScript. As long as the next loader in the chain can handle its output, the loader can return any type of module.
 
-A loader should be independent of previous compilations of the same module.
+### Modular
 
-### Use the [loader-utils](https://github.com/webpack/loader-utils)
+Keep the output modular. Loader generated modules should respect the same design principles as normal modules.
 
-In order to make the experience consistent for other developers, you should use the loader-utils to get the loader options:
+### Stateless
 
-```javascript
-const loaderUtils = require("loader-utils");
+Make sure the loader does not retain state between module transformations. Each run should always be independent of other compiled modules as well as previous compilations of the same module.
+
+### Loader Utilities
+
+Take advantage of the [`loader-utils`](https://github.com/webpack/loader-utils) package. It provides a variety of useful tools but one of the most common is retrieving the options passed to the loader:
+
+``` js
+const loaderUtils = require('loader-utils');
 
 module.exports = function(source) {
-    const options = loaderUtils.getOptions(this);
+  const options = loaderUtils.getOptions(this);
 };
 ```
 
-There are also other utility functions like `interpolateName`.
+### Loader Dependencies
 
-### Mark dependencies
+If a loader uses external resources (i.e. by reading from filesystem), they __must__ indicate it. This information is used to invalidate cacheable loaders and recompile in watch mode. Here's a brief example of how to accomplish this using the `addDependency` method:
 
-If a loader uses external resources (i. e. by reading from filesystem), they **must** tell about that. This information is used to invalidate cacheable loaders and recompile in watch mode.
+``` js
+const path = require('path');
 
-``` javascript
-// Loader adding a header
-var path = require("path");
 module.exports = function(source) {
   var callback = this.async();
-  var headerPath = path.resolve("header.js");
+  var headerPath = path.resolve('header.js');
+
   this.addDependency(headerPath);
-  fs.readFile(headerPath, "utf-8", function(err, header) {
+
+  fs.readFile(headerPath, 'utf-8', function(err, header) {
```

Comment:
```js
this.emitFile(file)
```

[LoaderAPI#.emitFile](https://webpack.js.org/api/loaders/#this-emitfile)

Reply:
Also good to know, thanks.

Reply:
Actually I'm a little confused on this one... would `emitFile` replace `addDependency`or `fs.readFile`? Can you elaborate?

Reply:
`fs.readFile` But my bad here, it was `readFile` sry 😛 , still `this.fs` could be used then, using the buildin fs (`compiler.inputFileSystem/compiler.outputFileSystem`)(`memory-fs`)

Reply:
Cool, will update to `this.fs.readFile`.